src/Mono.Android/Xamarin.Android.Net/X509TrustManagerWithValidationCallback.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
+
+using Javax.Net.Ssl;
+
+using JavaCertificateException = Java.Security.Cert.CertificateException;
+using JavaX509Certificate = Java.Security.Cert.X509Certificate;
+
+namespace Xamarin.Android.Net
+{
+	internal sealed class X509TrustManagerWithValidationCallback : Java.Lang.Object, IX509TrustManager
+	{
+		internal sealed class Helper
+		{
+			public Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> Callback { get; }
+
+			public Helper (Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> callback)
+			{
+				Callback = callback;
+			}
+
+			public ITrustManager[] Inject (
+				ITrustManager[]? trustManagers,
+				HttpRequestMessage requestMessage)
+			{
+				IX509TrustManager? x509TrustManager = trustManagers?.OfType<IX509TrustManager> ().FirstOrDefault ();
+				IEnumerable<ITrustManager> otherTrustManagers = trustManagers?.Where (manager => manager != x509TrustManager) ?? Enumerable.Empty<ITrustManager> ();
+				var trustManagerWithCallback = new X509TrustManagerWithValidationCallback (x509TrustManager, requestMessage, Callback);
+				return otherTrustManagers.Prepend (trustManagerWithCallback).ToArray ();
+			}
+		}
+
+		private readonly IX509TrustManager? _internalTrustManager;
+		private readonly HttpRequestMessage _request;
+		private readonly Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> _serverCertificateCustomValidationCallback;
+
+		private X509TrustManagerWithValidationCallback (
+			IX509TrustManager? internalTrustManager,
+			HttpRequestMessage request,
+			Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> serverCertificateCustomValidationCallback)
+		{
+			_request = request;
+			_internalTrustManager = internalTrustManager;
+			_serverCertificateCustomValidationCallback = serverCertificateCustomValidationCallback;
+		}
+
+		public void CheckServerTrusted (JavaX509Certificate[] javaChain, string authType)
+		{
+			var sslPolicyErrors = SslPolicyErrors.None;
+			var certificates = ConvertCertificates (javaChain);
+
+			try {
+				_internalTrustManager?.CheckServerTrusted (javaChain, authType);
+			} catch (JavaCertificateException) {
```

Comment:
I feel we should log this exception, not the whole trace, just the message, using the `Debug` severity so that it's not shown by default.